modules/reactor/src/meshgenerators/PeripheralRingMeshGenerator.C

```diff
@@ -162,31 +230,171 @@ PeripheralRingMeshGenerator::generate()
     points_array[0][i] = std::get<1>(azi_points[i]);
+    // Define outer layer of inner boundary layer
+    if (_peripheral_inner_boundary_layer_intervals)
+    {
+      // Outside point of the inner boundary layer
+      const Point ref_inner_boundary_shift =
+          (_peripheral_inner_boundary_layer_width / sin(input_bdry_angles[i])) * input_bdry_nd[i];
+      ref_inner_bdry_surf.push_back(points_array[0][i] + ref_inner_boundary_shift);
+      ref_inner_bdry_azi.push_back(atan2(ref_inner_bdry_surf.back()(1) - origin_y,
+                                         ref_inner_bdry_surf.back()(0) - origin_x));
+    }
+  }
+  // Check if any azimuthal angles are messed after inner boundary layer addition
+  std::vector<bool> delete_mark(input_ext_node_num, false);
+  std::vector<Real> interp_azi_data, interp_x_data, interp_y_data;
+  std::unique_ptr<LinearInterpolation> linterp_x, linterp_y;
+  if (_peripheral_inner_boundary_layer_intervals)
+  {
+    // Mark the to-be-deleted elements
+    for (unsigned int i = 0; i < input_ext_node_num; ++i)
+    {
+      if (!MooseUtils::absoluteFuzzyEqual(input_bdry_angles[i], M_PI / 2.0))
+      {
+        if (((ref_inner_bdry_surf[(i - 1) % input_ext_node_num] - origin_pt)
+                 .cross(ref_inner_bdry_surf[i] - origin_pt))(2) <= 0.0)
+          delete_mark[(i - 1) % input_ext_node_num] = true;
+        if (((ref_inner_bdry_surf[(i + 1) % input_ext_node_num] - origin_pt)
+                 .cross(ref_inner_bdry_surf[i] - origin_pt))(2) >= 0.0)
+          delete_mark[(i + 1) % input_ext_node_num] = true;
+      }
+    }
+    // Create vectors for interpolation
+    for (unsigned int i = 0; i < input_ext_node_num; ++i)
+    {
+      if (!delete_mark[i])
+      {
+        interp_azi_data.push_back(ref_inner_bdry_azi[i]);
+        interp_x_data.push_back(ref_inner_bdry_surf[i](0));
+        interp_y_data.push_back(ref_inner_bdry_surf[i](1));
+        if (interp_azi_data.size() > 1)
+          if (interp_azi_data.back() < interp_azi_data[interp_azi_data.size() - 2])
+            interp_azi_data.back() += 2 * M_PI;
+      }
+    }
+    const Real interp_x0 = interp_x_data.front();
+    const Real interp_xt = interp_x_data.back();
+    const Real interp_y0 = interp_y_data.front();
+    const Real interp_yt = interp_y_data.back();
+    if (interp_azi_data.front() > -M_PI)
+    {
+      interp_azi_data.insert(interp_azi_data.begin(), -M_PI);
+      interp_x_data.insert(interp_x_data.begin(), interp_xt);
+      interp_y_data.insert(interp_y_data.begin(), interp_yt);
+    }
+    if (interp_azi_data.back() < M_PI)
+    {
+      interp_azi_data.push_back(M_PI);
+      interp_x_data.push_back(interp_x0);
+      interp_y_data.push_back(interp_y0);
+    }
+    // Establish interpolation
+    linterp_x = libmesh_make_unique<LinearInterpolation>(interp_azi_data, interp_x_data);
+    linterp_y = libmesh_make_unique<LinearInterpolation>(interp_azi_data, interp_y_data);

```

Comment:
X and Y need to be interpolated. I want to keep the azimuthal angles unchanged for the inner boundary layer (otherwise the mesh either look weird or messed). Thus, I use normal surface direction to define the location of the outer boundary of the boundary layer (by defining key Point positions). These key positions do not have the original azimuthal angles, so they are just used as interpolation input.